// e2e-dashboard/parser/index.js
/**
 * Parser orchestrator — builds a complete run object from CLI inputs.
 * Coordinates Playwright parser, Maestro parser, and screenshot collection.
 */

import { readFileSync, existsSync } from 'node:fs';
import { join } from 'node:path';
import { parsePlaywright } from './playwright.js';
import { parseMaestro } from './maestro.js';
import { findScreenshots, copyScreenshots, findBestMaestroScreenshot } from './screenshots.js';

export async function buildRun(opts) {
  const {
    runId,
    sha = null,
    message = null,
    branch = null,
    playwrightPath = null,
    maestroAndroidPath = null,
    maestroIosPath = null,
    screenshotsDir = null,
    outputDir = null,
  } = opts;

  const suites = {};
  const screenshotFiles = screenshotsDir ? findScreenshots(screenshotsDir) : [];
  const screenshotMappings = [];

  // Parse Playwright
  if (playwrightPath) {
    const suite = parsePlaywrightSuite(playwrightPath, screenshotFiles, screenshotsDir, runId);
    if (suite) {
      suites.web = suite;
      screenshotMappings.push(...collectScreenshotMappings(suite.tests));
    }
  }

  // Parse Maestro Android
  if (maestroAndroidPath) {
    const suite = parseMaestroSuite(maestroAndroidPath, 'android', screenshotFiles, runId);
    if (suite) {
      suites.android = suite;
      screenshotMappings.push(...collectScreenshotMappings(suite.tests));
    }
  }

  // Parse Maestro iOS
  if (maestroIosPath) {
    const suite = parseMaestroSuite(maestroIosPath, 'ios', screenshotFiles, runId);
    if (suite) {
      suites.ios = suite;
      screenshotMappings.push(...collectScreenshotMappings(suite.tests));
    }
  }

  // Copy screenshots to output directory
  if (outputDir && screenshotMappings.length > 0) {
    await copyScreenshots(screenshotMappings, outputDir);
  }

  return {
    id: runId,
    sha,
    message: message ? message.slice(0, 120) : null,
    branch,
    timestamp: new Date().toISOString(),
    suites,
  };
}

function parsePlaywrightSuite(filePath, screenshotFiles, screenshotsDir, runId) {
  try {
    const json = JSON.parse(readFileSync(filePath, 'utf8'));
    const rawTests = parsePlaywright(json);
    const tests = rawTests.map(t => {
      const screenshot = resolveScreenshotPath(t.screenshotPath, screenshotFiles, screenshotsDir, 'web', t.name, runId);
      const failureScreenshot = resolveScreenshotPath(t.failureScreenshotPath, screenshotFiles, screenshotsDir, 'web', t.name, runId, '-failure');

      return {
        name: t.name,
        status: t.status,
        duration_ms: t.duration_ms,
        screenshot: screenshot?.outputRef || null,
        failureScreenshot: failureScreenshot?.outputRef || null,
        error: t.error,
        _screenshotSource: screenshot?.sourcePath || null,
        _outputName: screenshot?.outputName || null,
        _failureScreenshotSource: failureScreenshot?.sourcePath || null,
        _failureOutputName: failureScreenshot?.outputName || null,
      };
    });

    return buildSuiteObject('playwright', tests);
  } catch (err) {
    console.warn(`Warning: Could not parse Playwright report at ${filePath}: ${err.message}`);
    return null;
  }
}

function parseMaestroSuite(filePath, platform, screenshotFiles, runId) {
  try {
    const xml = readFileSync(filePath, 'utf8');
    const rawTests = parseMaestro(xml);
    const tests = rawTests.map(t => {
      // Match screenshot by flow name (non-failure for primary, failure for failureScreenshot)
      const matchedFile = findBestMaestroScreenshot(screenshotFiles, t.name, false);
      const failureFile = t.status === 'failed'
        ? findBestMaestroScreenshot(screenshotFiles, t.name, true)
        : null;

      const outputName = matchedFile ? `${platform}-${slugify(t.name)}.png` : null;
      const outputRef = outputName ? `screenshots/run-${runId}/${outputName}` : null;
      const failOutputName = failureFile ? `${platform}-${slugify(t.name)}-failure.png` : null;
      const failOutputRef = failOutputName ? `screenshots/run-${runId}/${failOutputName}` : null;

      return {
        name: t.name,
        status: t.status,
        duration_ms: t.duration_ms,
        screenshot: outputRef,
        failureScreenshot: failOutputRef,
        error: t.error,
        _screenshotSource: matchedFile || null,
        _outputName: outputName,
        _failureScreenshotSource: failureFile || null,
        _failureOutputName: failOutputName,
      };
    });

    return buildSuiteObject('maestro', tests);
  } catch (err) {
    console.warn(`Warning: Could not parse Maestro report at ${filePath}: ${err.message}`);
    return null;
  }
}

function resolveScreenshotPath(rawPath, screenshotFiles, screenshotsDir, prefix, testName, runId, suffix = '') {
  if (!rawPath) return null;

  const outputName = `${prefix}-${slugify(testName)}${suffix}.png`;
  const outputRef = `screenshots/run-${runId}/${outputName}`;

  // Try to find the actual file
  let sourcePath = null;
  if (screenshotsDir) {
    // Try the path as-is first (might be absolute)
    if (existsSync(rawPath)) {
      sourcePath = rawPath;
    }
    // Try relative to screenshots dir
    const relPath = join(screenshotsDir, rawPath);
    if (!sourcePath && existsSync(relPath)) {
      sourcePath = relPath;
    }
    // Fallback: scan for matching file
    if (!sourcePath) {
      const basename = rawPath.split('/').pop();
      sourcePath = screenshotFiles.find(f => f.endsWith(basename)) || null;
    }
  }

  if (!sourcePath) return null;
  return { sourcePath, outputName, outputRef };
}

function collectScreenshotMappings(tests) {
  const mappings = [];
  for (const test of tests) {
    if (test._screenshotSource) {
      mappings.push({ sourcePath: test._screenshotSource, outputName: test._outputName });
    }
    if (test._failureScreenshotSource) {
      mappings.push({ sourcePath: test._failureScreenshotSource, outputName: test._failureOutputName });
    }
    delete test._screenshotSource;
    delete test._outputName;
    delete test._failureScreenshotSource;
    delete test._failureOutputName;
  }
  return mappings;
}

function buildSuiteObject(framework, tests) {
  const passed = tests.filter(t => t.status === 'passed').length;
  const failed = tests.filter(t => t.status === 'failed').length;
  const skipped = tests.filter(t => t.status === 'skipped').length;
  const duration_ms = tests.reduce((sum, t) => sum + t.duration_ms, 0);

  return { framework, passed, failed, skipped, duration_ms, tests };
}

function slugify(name) {
  return name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
}
