// e2e-dashboard/parser/playwright.js
/**
 * Parse Playwright JSON report into a flat array of test results.
 * Recursively walks nested suites to extract all specs.
 */

const STATUS_MAP = {
  passed: 'passed',
  failed: 'failed',
  timedOut: 'failed',
  interrupted: 'failed',
  skipped: 'skipped',
};

export function parsePlaywright(report) {
  const tests = [];
  collectSpecs(report.suites || [], tests);
  return tests;
}

function collectSpecs(suites, tests) {
  for (const suite of suites) {
    for (const spec of (suite.specs || [])) {
      for (const test of (spec.tests || [])) {
        tests.push(parseTest(spec.title, test));
      }
    }
    collectSpecs(suite.suites || [], tests);
  }
}

function parseTest(name, test) {
  const result = test.results?.[0];

  if (!result || test.status === 'skipped') {
    return {
      name,
      status: 'skipped',
      duration_ms: 0,
      screenshotPath: null,
      failureScreenshotPath: null,
      error: null,
    };
  }

  const status = STATUS_MAP[result.status] || 'failed';
  const attachments = result.attachments || [];
  const pngAttachments = attachments.filter(a => a.contentType === 'image/png');

  const screenshot = pngAttachments.find(a => a.name === 'screenshot');
  const failureScreenshot = pngAttachments.find(a => a.name === 'screenshot-on-failure');

  const errorMsg = result.errors?.[0]?.message || null;

  return {
    name,
    status,
    duration_ms: result.duration || 0,
    screenshotPath: screenshot?.path || null,
    failureScreenshotPath: failureScreenshot?.path || null,
    error: errorMsg,
  };
}
