// e2e-dashboard/parser/screenshots.js
/**
 * Screenshot collection utilities.
 * Recursively scans directories for PNG files, copies them to output.
 * Optional sharp-based resizing for large screenshots.
 */

import { readdirSync, statSync, copyFileSync, mkdirSync, existsSync } from 'node:fs';
import { join, extname } from 'node:path';

/**
 * Recursively find all PNG files in a directory.
 * @param {string} dir - Directory to scan
 * @returns {string[]} Array of absolute paths to PNG files
 */
export function findScreenshots(dir) {
  if (!existsSync(dir)) return [];

  const results = [];
  walk(dir, results);
  return results;
}

function walk(dir, results) {
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      walk(fullPath, results);
    } else if (entry.isFile() && extname(entry.name).toLowerCase() === '.png') {
      results.push(fullPath);
    }
  }
}

/**
 * Copy screenshot files to output directory.
 * If sharp is available, resizes screenshots > 500KB.
 * @param {Array<{sourcePath: string, outputName: string}>} mapping
 * @param {string} outputDir
 */
export async function copyScreenshots(mapping, outputDir) {
  if (mapping.length === 0) return;
  mkdirSync(outputDir, { recursive: true });

  for (const { sourcePath, outputName } of mapping) {
    if (!existsSync(sourcePath)) {
      console.warn(`Screenshot not found, skipping: ${sourcePath}`);
      continue;
    }

    const destPath = join(outputDir, outputName);
    const fileSize = statSync(sourcePath).size;

    if (fileSize > 500_000) {
      const resized = await resizeScreenshot(sourcePath, destPath);
      if (!resized) {
        copyFileSync(sourcePath, destPath);
      }
    } else {
      copyFileSync(sourcePath, destPath);
    }
  }
}

/**
 * Resize a screenshot using sharp (optional peer dep).
 * Reduces dimensions by 50% iteratively until under 500KB.
 * @param {string} sourcePath
 * @param {string} destPath
 * @returns {Promise<boolean>} true if resized, false if sharp unavailable
 */
export async function resizeScreenshot(sourcePath, destPath) {
  try {
    const sharp = (await import('sharp')).default;
    const metadata = await sharp(sourcePath).metadata();
    let width = metadata.width;
    let height = metadata.height;

    while (width > 100) {
      width = Math.round(width / 2);
      height = Math.round(height / 2);
      const buffer = await sharp(sourcePath).resize(width, height).png().toBuffer();
      if (buffer.length <= 500_000) {
        await sharp(sourcePath).resize(width, height).png().toFile(destPath);
        return true;
      }
    }
    // Even at minimum size, write it out
    await sharp(sourcePath).resize(width, height).png().toFile(destPath);
    return true;
  } catch {
    return false;
  }
}

/**
 * Match a screenshot filename to a Maestro flow name.
 * Normalizes both names: lowercase, hyphens/underscores treated as equivalent.
 * @param {string} screenshotPath
 * @param {string} flowName
 * @returns {boolean}
 */
export function matchesMaestroFlow(screenshotPath, flowName) {
  const normalize = s => s.toLowerCase().replace(/[_-]/g, '');
  const filename = screenshotPath.split('/').pop().replace(/\.png$/i, '');
  return normalize(filename).includes(normalize(flowName));
}

/**
 * Check if a screenshot filename indicates a failure screenshot.
 * Maestro failure screenshots contain "fail" or "error" in the filename.
 * @param {string} screenshotPath
 * @returns {boolean}
 */
export function isFailureScreenshot(screenshotPath) {
  const filename = screenshotPath.split('/').pop().toLowerCase();
  return filename.includes('fail') || filename.includes('error');
}

/**
 * Find the best matching screenshot for a Maestro flow from a list of files.
 * If multiple match, returns the last-modified one.
 * @param {string[]} screenshotFiles - All available screenshot paths
 * @param {string} flowName - Maestro flow name to match
 * @param {boolean} failure - If true, match only failure screenshots
 * @returns {string|null} Best matching screenshot path
 */
export function findBestMaestroScreenshot(screenshotFiles, flowName, failure = false) {
  const matches = screenshotFiles.filter(f => {
    if (!matchesMaestroFlow(f, flowName)) return false;
    return failure ? isFailureScreenshot(f) : !isFailureScreenshot(f);
  });

  if (matches.length === 0) return null;
  if (matches.length === 1) return matches[0];

  // Return last-modified
  return matches.sort((a, b) => {
    const aStat = statSync(a);
    const bStat = statSync(b);
    return bStat.mtimeMs - aStat.mtimeMs;
  })[0];
}
